src/node.cc

```diff
@@ -3226,6 +3239,35 @@ static void SignalExit(int signo) {
   raise(signo);
 }
 
+#ifdef __POSIX__
+// Callbacks for triggering NodeReport on external signal - these run on the main event loop thread
+static void SignalDumpInterruptCallback(Isolate *isolate, void *data) {
+  if (nodereport_signal != 0) {
+    TriggerNodeReport(Isolate::GetCurrent(), kSignal_JS, signo_string(*((int *)data)), "node::SignalDumpInterruptCallback()");
+    nodereport_signal = 0;
+  }
+}
+static void SignalDumpAsyncCallback(uv_async_t* handle) {
+  if (nodereport_signal != 0) {
+    TriggerNodeReport(Isolate::GetCurrent(), kSignal_UV, signo_string((int)(size_t)handle->data), "node::SignalDumpAsyncCallback()");
+    nodereport_signal = 0;
+  }
+}
+
+// Raw signal handler for triggering a NodeReport - runs on an arbitrary thread
+static void SignalDump(int signo) {
+  // Check atomic for NodeReport already pending
+  if (__sync_val_compare_and_swap(&nodereport_signal, 0, signo) == 0) {
+    fprintf(stderr,"Signal %s received, triggering NodeReport\n", signo_string(signo));
+    Isolate::GetCurrent()->RequestInterrupt(SignalDumpInterruptCallback, &nodereport_signal);
```

Comment:
OK, I will need to re-think this signal handler. Maybe hand over to a dedicated thread that's waiting on a semaphore, and call the RequestInterrupt() and uv_async_send() APIs from there.